test/box-luatest/mvcc_ddl_test.lua

```diff
@@ -0,0 +1,467 @@
+
+-- The test covers a crash when transaction that is being deleted removes
+-- itself from reader list of a deleted story that leads to use-after-free
+g.test_reader_list_use_after_free = function(cg)
+    cg.server:exec(function()
+        local txn_proxy = require("test.box.lua.txn_proxy")
+
+        -- Create space with tuples
+        local s = box.schema.space.create('test')
+        s:create_index('pk')
+
+        box.begin()
+        for i = 1, 10000 do
+            s:replace{i}
+        end
+        box.commit()
+
+        -- Create a transaction that reads every tuple so it's
+        -- inserted to reader list of every story
+        local tx = txn_proxy.new()
+        tx:begin()
+        for i = 1, 10000 do
+            tx('box.space.test:get{' .. i .. '}')
+        end
+
+        -- Create a new index
+        -- Firstly, we need it so that all the stories will be deleted
+        -- due to DDL
+        -- Secondly, we need to create a new index so that layout of stories
+        -- will be changed and use-after-free on rlist link will trash another
+        -- field (for example, pointer to tuple) and that's will definitely lead
+        -- to crash
+        box.space.test:create_index('sk')
+
+        -- Open a read-view so that stories for all tuples from the space
+        -- are created
+        local rv = txn_proxy.new()
+        rv:begin()
+        rv('box.space.test:select{}')
+
+        -- Rollback the first reader so that it will delete itself from reader
+        -- lists of all stories and that will lead to use-after-free
+        tx:rollback()
+
+        -- Read all the tuples, Tarantool is most likely to crash here if
+        -- use-after-free broke something
+        for i = 1, 10000 do
+            s:get{i}
+        end
+    end)
+end
+
+-- The test checks if stories associated with old schema cannot
+-- be access on index creation
+-- gh-10096
+g.test_dont_access_old_stories_on_index_create = function(cg)
+    cg.server:exec(function()
+        local fiber = require('fiber')
+
+        local s = box.schema.space.create('test')
+        s:create_index('pk', {parts = {{1}}})
+        s:create_index('sk1', {parts = {{2}}, unique = true})
+
+        s:insert{1, 1, 1, 0}
+        -- Collect stories to be independent from GC steps
+        box.internal.memtx_tx_gc(100)
+
+        -- Create index in a separate transaction
+        -- Transaction creating index will create story for our tuple
+        -- when it reads it to insert into the new index
+        -- The problem is the story is created with old index_count and
+        -- does not have links for created index
+        local created_index = false
+        local f = fiber.create(function()
+            s:create_index('sk2', {parts = {{3}}, unique = true})
+            created_index = true
+        end)
+        f:set_joinable(true)
+
+        -- Make sure txn is still in progress
+        assert(not created_index)
+        s:replace{1, 1, 1, 1}
+        f:join()
```

Comment:
Is the join supposed to fail or succeed? Please add an assertion for the join and go over the rest of the `fiber.join`s in the test where assertions are missing (there is a bunch).

Reply:
Added check to all fiber joins.